libarchive/archive_read_disk_posix.c

```diff
@@ -828,7 +832,15 @@ _archive_read_data_block(struct archive *_a, const void **buff,
 		}
 	} else
 		bytes = 0;
-	if (bytes == 0) {
+	/*
+	 * Return an EOF unless we've read a leading empty sparse region, which
+	 * is used to represent fully-sparse files.
+	 *
+	 * TODO: it is not technically necessary to check for entry_tota == 0,
+	 * but this simplifies some unit tests that expect to only read data
+	 * regions with length > 0. Consider fixing these tests (test_sparse_basic).
+	*/
+	if (bytes == 0 && !(empty_sparse_region && t->entry_total == 0)) {
```

Comment:
Fails for me with the following diff:

```
-       if (bytes == 0 && !(empty_sparse_region && t->entry_total == 0)) {
+       if (bytes == 0 && !(empty_sparse_region)) {
```

Failure has the following assertion:

```
test_sparse_basic

/devel/libarchive/libarchive/test/test_sparse_basic.c:291: expected_data_blocks != data_blocks
      expected_data_blocks=2 (0x2, 02)
      data_blocks=3 (0x3, 03)
```

I'm not sure what would cause you to not see this failure.